File: src/email_parser/workflow/state.py
#!/usr/bin/env python3
"""
Workflow State Schema

Defines the state structure for LangGraph workflow execution.
The state evolves as it flows through workflow nodes.

Reference: docs/IAB_TAXONOMY_PROFILE_REQUIREMENTS.md - LangGraph Workflow section
"""

from typing import TypedDict, List, Dict, Any, Optional
from typing_extensions import NotRequired


class WorkflowState(TypedDict):
    """
    State schema for IAB Taxonomy Profile workflow.

    This state is passed between workflow nodes and evolves as the
    workflow progresses through different stages.

    State Flow:
    1. load_emails → adds emails, processed_email_ids
    2. retrieve_profile → adds existing_profile
    3. analyzers → add analysis results
    4. reconcile → adds reconciliation_data
    5. update_memory → adds updated_profile

    Fields are marked as Required or NotRequired based on when they
    become available in the workflow.
    """

    # =============================================================================
    # User Context (Required from start)
    # =============================================================================

    user_id: str
    """Unique user identifier for memory isolation."""

    # =============================================================================
    # Email Data (Added by load_emails node)
    # =============================================================================

    emails: NotRequired[List[Dict[str, Any]]]
    """
    List of new emails to process.

    Each email dict contains:
    - id: str (email identifier)
    - subject: str
    - body: str
    - from_addr: str
    - to_addr: str
    - date: str (ISO 8601)
    - ... other email fields
    """

    processed_email_ids: NotRequired[List[str]]
    """Email IDs that have already been processed (from memory)."""

    current_email_index: NotRequired[int]
    """Index of current email being processed (for iteration - DEPRECATED, use batch fields)."""

    total_emails: NotRequired[int]
    """Total number of emails to process."""

    # =============================================================================
    # Batch Processing Fields (Added for performance optimization)
    # =============================================================================

    current_batch_start: NotRequired[int]
    """Starting index of current batch in emails list."""

    batch_size: NotRequired[int]
    """Number of emails in current batch."""

    model_context_window: NotRequired[int]
    """Context window size (tokens) of the LLM model being used."""

    force_reprocess: NotRequired[bool]
    """If True, ignore processed email tracking and reprocess all emails."""

    cost_tracker: NotRequired[Any]
    """CostTracker instance for tracking LLM API costs (optional)."""

    tracker: NotRequired[Any]
    """WorkflowTracker instance for dashboard analytics tracking (optional)."""

    llm_model: NotRequired[str]
    """LLM model to use for IAB taxonomy classification (optional)."""

    # =============================================================================
    # Profile Data (Added by retrieve_profile node)
    # =============================================================================

    existing_profile: NotRequired[Dict[str, Any]]
    """
    Current taxonomy profile from memory (before updates).

    Structure:
    {
        "demographics": [SemanticMemory, ...],
        "household": [SemanticMemory, ...],
        "interests": [SemanticMemory, ...],
        "purchase_intent": [SemanticMemory, ...],
        "actual_purchases": [SemanticMemory, ...]
    }
    """

    # =============================================================================
    # Analysis Results (Added by analyzer nodes)
    # =============================================================================

    demographics_results: NotRequired[List[Dict[str, Any]]]
    """
    Taxonomy selections from demographics analyzer.

    Each selection:
    {
        "taxonomy_id": int,
        "section": "demographics",
        "value": str,
        "confidence": float,
        "category_path": str,
        "tier_1": str,
        "tier_2": str,
        "tier_3": str,
        "tier_4": str,
        "tier_5": str,
        "reasoning": str
    }
    """

    household_results: NotRequired[List[Dict[str, Any]]]
    """Taxonomy selections from household analyzer."""

    interests_results: NotRequired[List[Dict[str, Any]]]
    """Taxonomy selections from interests analyzer."""

    purchase_results: NotRequired[List[Dict[str, Any]]]
    """Taxonomy selections from purchase analyzer."""

    # =============================================================================
    # Reconciliation Data (Added by reconcile node)
    # =============================================================================

    reconciliation_data: NotRequired[List[Dict[str, Any]]]
    """
    Results of evidence reconciliation.

    List of updated semantic memories after reconciliation.
    """

    # =============================================================================
    # Output (Added by update_memory node)
    # =============================================================================

    updated_profile: NotRequired[Dict[str, Any]]
    """
    Final updated profile after memory consolidation.

    Structure matches existing_profile but with updated values.
    """

    # =============================================================================
    # Metadata & Error Tracking
    # =============================================================================

    errors: NotRequired[List[str]]
    """List of error messages encountered during workflow execution."""

    warnings: NotRequired[List[str]]
    """List of warning messages (non-fatal issues)."""

    workflow_started_at: NotRequired[str]
    """ISO 8601 timestamp when workflow started."""

    workflow_completed_at: NotRequired[str]
    """ISO 8601 timestamp when workflow completed."""

    # =============================================================================
    # Routing Metadata (Used for conditional edges)
    # =============================================================================

    next_analyzers: NotRequired[List[str]]
    """
    List of analyzer node names to execute for current email.

    Set by routing logic based on email content.
    Examples: ["demographics", "interests"], ["purchase"]
    """

    completed_analyzers: NotRequired[List[str]]
    """Analyzer nodes that have completed for current email."""


# =============================================================================
# Helper Functions for State Initialization
# =============================================================================

def create_initial_state(
    user_id: str,
    emails: List[Dict[str, Any]],
    force_reprocess: bool = False,
    model_context_window: Optional[int] = None
) -> WorkflowState:
    """
    Create initial workflow state.

    Args:
        user_id: User identifier
        emails: List of email dictionaries to process
        force_reprocess: If True, ignore processed email tracking
        model_context_window: Context window size of LLM model (for batch sizing)

    Returns:
        Initial WorkflowState with required fields populated

    Example:
        >>> state = create_initial_state("user_123", emails, model_context_window=128000)
        >>> workflow.invoke(state)
    """
    # Import batch_optimizer here to avoid circular dependency
    from ..workflow.batch_optimizer import calculate_batch_size

    # Calculate initial batch size if context window provided
    initial_batch_size = 1  # Default to single-email processing
    if model_context_window:
        initial_batch_size = calculate_batch_size(
            emails=emails,
            context_window=model_context_window,
            start_index=0
        )

    return WorkflowState(
        user_id=user_id,
        emails=emails,
        processed_email_ids=[],
        current_email_index=0,  # Deprecated, use batch fields
        total_emails=len(emails),
        force_reprocess=force_reprocess,
        # Batch processing fields
        current_batch_start=0,
        batch_size=initial_batch_size,
        model_context_window=model_context_window,
        # Results
        demographics_results=[],
        household_results=[],
        interests_results=[],
        purchase_results=[],
        reconciliation_data=[],
        errors=[],
        warnings=[],
        next_analyzers=[],
        completed_analyzers=[],
    )


def get_current_email(state: WorkflowState) -> Optional[Dict[str, Any]]:
    """
    Get the current email being processed.

    DEPRECATED: Use get_current_batch() for batch processing.

    Args:
        state: Current workflow state

    Returns:
        Current email dictionary or None if no emails

    Example:
        >>> email = get_current_email(state)
        >>> if email:
        ...     print(email['subject'])
    """
    if "emails" not in state or not state["emails"]:
        return None

    index = state.get("current_email_index", 0)
    if index >= len(state["emails"]):
        return None

    return state["emails"][index]


def get_current_batch(state: WorkflowState) -> List[Dict[str, Any]]:
    """
    Get the current batch of emails being processed.

    Args:
        state: Current workflow state

    Returns:
        List of emails in current batch (empty list if no emails)

    Example:
        >>> batch = get_current_batch(state)
        >>> print(f"Processing {len(batch)} emails")
    """
    from ..workflow.batch_optimizer import get_batch_from_state
    return get_batch_from_state(state)


def has_more_emails(state: WorkflowState) -> bool:
    """
    Check if there are more emails to process.

    Supports both single-email (deprecated) and batch processing.

    Args:
        state: Current workflow state

    Returns:
        True if more emails remain, False otherwise

    Example:
        >>> if has_more_emails(state):
        ...     # Continue processing
        ... else:
        ...     # Workflow complete
    """
    if "emails" not in state or not state["emails"]:
        return False

    # Use batch fields if available, fall back to single-email index
    if "current_batch_start" in state:
        from ..workflow.batch_optimizer import has_more_batches
        return has_more_batches(state)
    else:
        # Legacy single-email processing
        current_index = state.get("current_email_index", 0)
        total = state.get("total_emails", 0)
        return current_index < total


def advance_to_next_email(state: WorkflowState) -> WorkflowState:
    """
    Advance state to next email or batch.

    Supports both single-email (deprecated) and batch processing.

    Args:
        state: Current workflow state

    Returns:
        Updated state with incremented index/batch and reset analyzer results

    Example:
        >>> state = advance_to_next_email(state)
        >>> # Now processing next email/batch
    """
    # Use batch advancement if batch fields present
    if "current_batch_start" in state:
        from ..workflow.batch_optimizer import advance_to_next_batch
        state = advance_to_next_batch(state)
    else:
        # Legacy single-email processing
        state["current_email_index"] = state.get("current_email_index", 0) + 1

    # Reset analyzer results for next email/batch
    state["demographics_results"] = []
    state["household_results"] = []
    state["interests_results"] = []
    state["purchase_results"] = []
    state["next_analyzers"] = []
    state["completed_analyzers"] = []

    return state


def add_error(state: WorkflowState, error_message: str) -> WorkflowState:
    """
    Add an error message to state.

    Args:
        state: Current workflow state
        error_message: Error description

    Returns:
        Updated state with error added
    """
    if "errors" not in state:
        state["errors"] = []

    state["errors"].append(error_message)
    return state


def add_warning(state: WorkflowState, warning_message: str) -> WorkflowState:
    """
    Add a warning message to state.

    Args:
        state: Current workflow state
        warning_message: Warning description

    Returns:
        Updated state with warning added
    """
    if "warnings" not in state:
        state["warnings"] = []

    state["warnings"].append(warning_message)
    return state

File: src/browser/agents/iab-classifier/state.ts
/**
 * IAB Classifier Workflow State
 *
 * TypeScript port of Python WorkflowState from:
 * src/email_parser/workflow/state.py (lines 15-404)
 *
 * This file provides:
 * - WorkflowState interface (LangGraph.js state schema)
 * - State initialization and manipulation helpers
 * - Batch processing utilities
 *
 * MIGRATION NOTE: This is an EXACT 1:1 translation of the Python implementation.
 * Every type, field, and function has been verified against the Python source.
 */

import { Annotation } from '@langchain/langgraph/web'
import {
  getBatchFromState,
  hasMoreBatches,
  advanceToNextBatch as batchOptimizerAdvance,
  calculateBatchSize,
} from './batchOptimizer'

/**
 * WorkflowStateInterface - TypeScript interface for type checking
 *
 * Python source: state.py:15-194 (class WorkflowState(TypedDict))
 *
 * This interface defines all state fields used throughout the classification workflow.
 * It serves as the type definition for the state schema.
 *
 * Note: The actual state schema object is created with Annotation.Root() below
 * and exported as WorkflowState for use with StateGraph.
 *
 * Type Mapping Legend:
 * - Python str → TypeScript string
 * - Python int → TypeScript number
 * - Python bool → TypeScript boolean
 * - Python List[T] → TypeScript Array<T>
 * - Python Dict[str, Any] → TypeScript Record<string, any>
 * - Python Optional[T] → TypeScript T | undefined
 * - Python NotRequired[T] → TypeScript T? (optional property)
 * - Python Any → TypeScript any
 */
export interface WorkflowStateInterface {
  // ============================================================================
  // REQUIRED FIELDS
  // ============================================================================

  /**
   * User ID
   * Python: state.py:37-38
   *   user_id: str
   *   """User ID for whom the profile is being built."""
   */
  user_id: string

  // ============================================================================
  // EMAIL DATA (Optional)
  // ============================================================================

  /**
   * List of email dictionaries to process
   * Python: state.py:44-56
   *   emails: NotRequired[List[Dict[str, Any]]]
   *   """List of email dictionaries to process.
   *
   *   Each email dict contains:
   *   - id: str
   *   - subject: str
   *   - body: str
   *   - sender: str
   *   - date: str
   *   - ... (other email metadata)
   *   """
   */
  emails?: Array<Record<string, any>>

  /**
   * List of email IDs that have been processed
   * Python: state.py:58-59
   *   processed_email_ids: NotRequired[List[str]]
   *   """List of email IDs that have been processed."""
   */
  processed_email_ids?: Array<string>

  /**
   * Current email index (DEPRECATED - use batch processing instead)
   * Python: state.py:61-62
   *   current_email_index: NotRequired[int]
   *   """Current index in the emails list (for single-email processing)."""
   */
  current_email_index?: number

  /**
   * Total number of emails
   * Python: state.py:64-65
   *   total_emails: NotRequired[int]
   *   """Total number of emails to process."""
   */
  total_emails?: number

  // ============================================================================
  // BATCH PROCESSING
  // ============================================================================

  /**
   * Starting index of current batch
   * Python: state.py:71-72
   *   current_batch_start: NotRequired[int]
   *   """Starting index of the current batch being processed."""
   */
  current_batch_start?: number

  /**
   * Number of emails in current batch
   * Python: state.py:74-75
   *   batch_size: NotRequired[int]
   *   """Number of emails to process in current batch."""
   */
  batch_size?: number

  /**
   * Model's context window size (tokens)
   * Python: state.py:77-78
   *   model_context_window: NotRequired[int]
   *   """Model's context window size in tokens."""
   */
  model_context_window?: number

  /**
   * Force reprocessing of already-processed emails
   * Python: state.py:80-81
   *   force_reprocess: NotRequired[bool]
   *   """Whether to reprocess emails that have already been processed."""
   */
  force_reprocess?: boolean

  // ============================================================================
  // LLM CONFIGURATION
  // ============================================================================

  /**
   * Cost tracker instance (for monitoring API costs)
   * Python: state.py:83-84
   *   cost_tracker: NotRequired[Any]
   *   """Cost tracker instance."""
   */
  cost_tracker?: any

  /**
   * Progress tracker instance
   * Python: state.py:86-87
   *   tracker: NotRequired[Any]
   *   """Progress tracker instance."""
   */
  tracker?: any

  /**
   * LLM model name
   * Python: state.py:89-90
   *   llm_model: NotRequired[str]
   *   """LLM model to use for classification."""
   */
  llm_model?: string

  /**
   * LLM configuration (API keys, etc.)
   * NOTE: For Phase 1.5 development only (admin dashboard).
   * In production (Phase 5), users provide their own keys via Settings UI.
   */
  llm_config?: {
    api_key?: string
    model?: string
    temperature?: number
    max_tokens?: number
    base_url?: string
  }

  /**
   * Injected LLM client (for testing)
   */
  llm_client?: any

  // ============================================================================
  // PROFILE DATA
  // ============================================================================

  /**
   * Existing IAB profile for the user
   * Python: state.py:96-106
   *   existing_profile: NotRequired[Dict[str, Any]]
   *   """Existing IAB profile retrieved from memory.
   *
   *   Structure:
   *   {
   *       "user_id": str,
   *       "demographics": List[Dict],
   *       "household": List[Dict],
   *       "interests": List[Dict],
   *       "purchase_intent": List[Dict]
   *   }
   *   """
   */
  existing_profile?: Record<string, any>

  // ============================================================================
  // ANALYSIS RESULTS (One per analyzer)
  // ============================================================================

  /**
   * Demographics analyzer results
   * Python: state.py:114-131
   *   demographics_results: NotRequired[List[Dict[str, Any]]]
   *   """Results from demographics analyzer.
   *
   *   Each result contains:
   *   - taxonomy_id: int
   *   - section: str
   *   - value: str
   *   - confidence: float
   *   - tier_1, tier_2, tier_3, tier_4, tier_5: str
   *   - category_path: str
   *   - grouping_tier_key: str
   *   - grouping_value: str
   *   - reasoning: str
   *   """
   */
  demographics_results?: Array<Record<string, any>>

  /**
   * Household analyzer results
   * Python: state.py:134-135
   *   household_results: NotRequired[List[Dict[str, Any]]]
   *   """Results from household analyzer."""
   */
  household_results?: Array<Record<string, any>>

  /**
   * Interests analyzer results
   * Python: state.py:137-138
   *   interests_results: NotRequired[List[Dict[str, Any]]]
   *   """Results from interests analyzer."""
   */
  interests_results?: Array<Record<string, any>>

  /**
   * Purchase intent analyzer results
   * Python: state.py:140-145
   *   purchase_results: NotRequired[List[Dict[str, Any]]]
   *   """Results from purchase intent analyzer.
   *
   *   Includes purchase_intent_flag field.
   *   """
   */
  purchase_results?: Array<Record<string, any>>

  // ============================================================================
  // RECONCILIATION
  // ============================================================================

  /**
   * Reconciliation data (merged classifications across analyzers)
   * Python: state.py:147-156
   *   reconciliation_data: NotRequired[List[Dict[str, Any]]]
   *   """Data from reconciliation process.
   *
   *   Contains merged classifications from all analyzers,
   *   with confidence scores adjusted based on:
   *   - Temporal decay
   *   - Evidence quality
   *   - Multi-source agreement
   *   """
   */
  reconciliation_data?: Array<Record<string, any>>

  /**
   * Updated profile after reconciliation
   * Python: state.py:158-167
   *   updated_profile: NotRequired[Dict[str, Any]]
   *   """Updated IAB profile after reconciliation.
   *
   *   Structure matches existing_profile:
   *   - demographics: List[Dict]
   *   - household: List[Dict]
   *   - interests: List[Dict]
   *   - purchase_intent: List[Dict]
   *   """
   */
  updated_profile?: Record<string, any>

  // ============================================================================
  // ERROR TRACKING
  // ============================================================================

  /**
   * List of error messages
   * Python: state.py:169-170
   *   errors: NotRequired[List[str]]
   *   """List of error messages encountered during workflow."""
   */
  errors?: Array<string>

  /**
   * List of warning messages
   * Python: state.py:172-173
   *   warnings: NotRequired[List[str]]
   *   """List of warning messages encountered during workflow."""
   */
  warnings?: Array<string>

  /**
   * Workflow start timestamp (ISO 8601)
   * Python: state.py:175-176
   *   workflow_started_at: NotRequired[str]
   *   """ISO 8601 timestamp when workflow started."""
   */
  workflow_started_at?: string

  /**
   * Workflow completion timestamp (ISO 8601)
   * Python: state.py:178-179
   *   workflow_completed_at: NotRequired[str]
   *   """ISO 8601 timestamp when workflow completed."""
   */
  workflow_completed_at?: string

  // ============================================================================
  // ROUTING (For conditional edges)
  // ============================================================================

  /**
   * List of analyzer node names to execute next
   * Python: state.py:185-191
   *   next_analyzers: NotRequired[List[str]]
   *   """List of analyzer node names to execute next.
   *
   *   Used by routing logic to determine which analyzers
   *   should run for the current email batch.
   *   """
   */
  next_analyzers?: Array<string>

  /**
   * List of analyzer node names that have completed
   * Python: state.py:193-194
   *   completed_analyzers: NotRequired[List[str]]
   *   """List of analyzer node names that have completed execution."""
   */
  completed_analyzers?: Array<string>
}

// ============================================================================
// STATE INITIALIZATION
// ============================================================================

/**
 * Create initial workflow state
 *
 * Python source: state.py:201-256
 *
 * @param user_id - User ID for whom to build profile
 * @param emails - List of email objects to process
 * @param force_reprocess - Whether to reprocess already-processed emails
 * @param model_context_window - LLM context window size (optional)
 * @returns Initial WorkflowState
 */
export function createInitialState(
  user_id: string,
  emails: Array<Record<string, any>>,
  force_reprocess: boolean = false,
  model_context_window?: number
): WorkflowStateInterface {
  // Python line 224: Import to avoid circular dependency
  // from ..workflow.batch_optimizer import calculate_batch_size

  // Python lines 227-233: Calculate batch size
  let initial_batch_size = 1
  if (model_context_window) {
    // Python lines 228-233:
    // initial_batch_size = calculate_batch_size(
    //     emails=emails,
    //     context_window=model_context_window,
    //     start_index=0
    // )
    initial_batch_size = calculateBatchSize(emails, model_context_window, 0)
  }

  // Python lines 235-256: Return WorkflowState
  return {
    user_id,
    emails,
    processed_email_ids: [],
    current_email_index: 0,
    total_emails: emails.length,
    force_reprocess,
    current_batch_start: 0,
    batch_size: initial_batch_size,
    model_context_window,
    demographics_results: [],
    household_results: [],
    interests_results: [],
    purchase_results: [],
    reconciliation_data: [],
    errors: [],
    warnings: [],
    next_analyzers: [],
    completed_analyzers: [],
  }
}

// ============================================================================
// STATE QUERY HELPERS
// ============================================================================

/**
 * Get current email from state (DEPRECATED - use get_current_batch instead)
 *
 * Python source: state.py:259-283
 *
 * @param state - Current workflow state
 * @returns Current email or undefined if no more emails
 */
export function getCurrentEmail(
  state: WorkflowStateInterface
): Record<string, any> | undefined {
  // Python lines 276-277: Check existence
  if (!state.emails || state.emails.length === 0) {
    return undefined
  }

  // Python lines 279-281: Bounds check
  const index = state.current_email_index ?? 0
  if (index >= state.emails.length) {
    return undefined
  }

  // Python line 283: Return
  return state.emails[index]
}

/**
 * Get current batch of emails from state
 *
 * Python source: state.py:286-301
 *
 * Delegates to batch_optimizer.get_batch_from_state()
 *
 * @param state - Current workflow state
 * @returns Array of emails in current batch
 */
export function getCurrentBatch(
  state: WorkflowStateInterface
): Array<Record<string, any>> {
  // Python lines 300-301: Delegate to batch_optimizer
  // from ..workflow.batch_optimizer import get_batch_from_state
  // return get_batch_from_state(state)
  return getBatchFromState(state)
}

/**
 * Check if there are more emails to process
 *
 * Python source: state.py:304-333
 *
 * @param state - Current workflow state
 * @returns True if more emails remain
 */
export function hasMoreEmails(state: WorkflowStateInterface): boolean {
  // Python lines 322-323: Check existence
  if (!state.emails || state.emails.length === 0) {
    return false
  }

  // Python lines 326-333: Batch vs single-email logic
  if (state.current_batch_start !== undefined) {
    // Batch processing mode
    // from ..workflow.batch_optimizer import has_more_batches
    // return has_more_batches(state)
    return hasMoreBatches(state)
  } else {
    // Single-email mode (DEPRECATED)
    const current_index = state.current_email_index ?? 0
    const total = state.total_emails ?? 0
    return current_index < total
  }
}

// ============================================================================
// STATE MUTATION HELPERS
// ============================================================================

/**
 * Advance to next email/batch
 *
 * Python source: state.py:336-368
 *
 * IMPORTANT: This function MUTATES the state object (Python pattern)
 * In TypeScript/LangGraph.js, we may want to return a new state object instead
 *
 * @param state - Current workflow state (will be mutated)
 * @returns Updated state
 */
export function advanceToNextEmail(state: WorkflowStateInterface): WorkflowStateInterface {
  // Python lines 353-358: Advance logic
  if (state.current_batch_start !== undefined) {
    // Batch processing mode
    // from ..workflow.batch_optimizer import advance_to_next_batch
    // state = advance_to_next_batch(state)
    batchOptimizerAdvance(state) // Mutates state in-place
  } else {
    // Single-email mode (DEPRECATED)
    state.current_email_index = (state.current_email_index ?? 0) + 1
  }

  // Python lines 361-366: Reset analyzer results
  state.demographics_results = []
  state.household_results = []
  state.interests_results = []
  state.purchase_results = []
  state.next_analyzers = []
  state.completed_analyzers = []

  // Python line 368: Return
  return state
}

/**
 * Add error message to state
 *
 * Python source: state.py:371-386
 *
 * @param state - Current workflow state (will be mutated)
 * @param error_message - Error message to add
 * @returns Updated state
 */
export function addError(
  state: typeof WorkflowState.State,
  error_message: string
): typeof WorkflowState.State {
  // Python lines 382-383: Initialize if needed
  if (!state.errors) {
    state.errors = []
  }

  // Python line 385: Append
  state.errors.push(error_message)
  return state
}

/**
 * Add warning message to state
 *
 * Python source: state.py:389-404
 *
 * @param state - Current workflow state (will be mutated)
 * @param warning_message - Warning message to add
 * @returns Updated state
 */
export function addWarning(
  state: typeof WorkflowState.State,
  warning_message: string
): typeof WorkflowState.State {
  // Python lines 400-401: Initialize if needed
  if (!state.warnings) {
    state.warnings = []
  }

  // Python line 403: Append
  state.warnings.push(warning_message)
  return state
}

/**
 * WorkflowState Annotation Schema for LangGraph.js
 *
 * LangGraph.js requires Annotation.Root() to create a state schema that can
 * be passed to StateGraph constructor.
 *
 * This creates the actual state schema object from the WorkflowState interface.
 */
export const WorkflowState = Annotation.Root({
  // Required fields
  user_id: Annotation<string>,

  // Email data
  emails: Annotation<Array<Record<string, any>>>({
    value: (_prev, current) => current,
    default: () => [],
  }),
  processed_email_ids: Annotation<Array<string>>({
    value: (_prev, current) => current,
    default: () => [],
  }),
  current_email_index: Annotation<number>({
    value: (_prev, current) => current,
    default: () => 0,
  }),
  total_emails: Annotation<number>({
    value: (_prev, current) => current,
    default: () => 0,
  }),

  // Existing profile
  existing_profile: Annotation<Record<string, any>>({
    value: (_prev, current) => current,
    default: () => ({}),
  }),

  // Analyzer results
  demographics_results: Annotation<Array<Record<string, any>>>({
    value: (_prev, current) => current,
    default: () => [],
  }),
  household_results: Annotation<Array<Record<string, any>>>({
    value: (_prev, current) => current,
    default: () => [],
  }),
  interests_results: Annotation<Array<Record<string, any>>>({
    value: (prev, current) => prev.concat(current),
    default: () => [],
  }),
  purchase_results: Annotation<Array<Record<string, any>>>({
    value: (prev, current) => prev.concat(current),
    default: () => [],
  }),

  // Reconciliation
  reconciliation_data: Annotation<Array<Record<string, any>>>({
    value: (_prev, current) => current,
    default: () => [],
  }),
  updated_profile: Annotation<Record<string, any>>({
    value: (_prev, current) => current,
    default: () => ({}),
  }),

  // Error tracking
  errors: Annotation<Array<string>>({
    value: (_prev, current) => current,
    default: () => [],
  }),
  warnings: Annotation<Array<string>>({
    value: (_prev, current) => current,
    default: () => [],
  }),

  // Timestamps
  workflow_started_at: Annotation<string>({
    value: (_prev, current) => current,
    default: () => '',
  }),
  workflow_completed_at: Annotation<string>({
    value: (_prev, current) => current,
    default: () => '',
  }),

  // Batch processing
  batch_size: Annotation<number>({
    value: (_prev, current) => current,
    default: () => 1,
  }),
  current_batch_index: Annotation<number>({
    value: (_prev, current) => current,
    default: () => 0,
  }),
  current_batch_start: Annotation<number>({
    value: (_prev, current) => current,
    default: () => 0,
  }),
  total_batches: Annotation<number>({
    value: (_prev, current) => current,
    default: () => 0,
  }),
  batch_mode: Annotation<boolean>({
    value: (_prev, current) => current,
    default: () => false,
  }),
  model_context_window: Annotation<number>({
    value: (_prev, current) => current,
    default: () => 0,
  }),
  force_reprocess: Annotation<boolean>({
    value: (_prev, current) => current,
    default: () => false,
  }),

  // LLM config
  cost_tracker: Annotation<any>({
    value: (_prev, current) => current,
    default: () => null,
  }),
  tracker: Annotation<any>({
    value: (_prev, current) => current,
    default: () => null,
  }),
  llm_provider: Annotation<string>({
    value: (_prev, current) => current,
    default: () => '',
  }),
  llm_model: Annotation<string>({
    value: (_prev, current) => current,
    default: () => '',
  }),
  llm_config: Annotation<any>({
    value: (_prev, current) => current,
    default: () => null,
  }),
  llm_client: Annotation<any>({
    value: (_prev, current) => current,
    default: () => null,
  }),

  // Routing (deprecated)
  next_analyzers: Annotation<Array<string>>({
    value: (_prev, current) => current,
    default: () => [],
  }),
  completed_analyzers: Annotation<Array<string>>({
    value: (_prev, current) => current,
    default: () => [],
  }),
})
